### ISSUE TITLE
CloudEvent API doesn't make sense

### ISSUE THREAD

**Issue by jroper** [lightbend/kalix-jvm-sdk#1024]
We always add CloudEvent information to outgoing events, including generating an id, setting the type, and setting the source. A user should be able to rely on that being set by us, and should be able to set additional cloud event metadata, without having to set that themselves. The problem is that right now, in order for a user to set CloudEvent metadata on an outgoing event using our API, they need to set the id, type and source.

There are three different methods we provide that allow them to get an instance of CloudEvent.

The first is `Metadata.asCloudEvent()`. This method will throw an exception if the underlying metadata doesn't contain a CloudEvent id, type or source. Since when publishing, you will typically start with `Metadata.EMPTY`, this is going to be the case.

The second is `Metadata.asCloudEvent(String, String, String)`. This requires explicitly passing an id, type or source.

The third is `CloudEvent.of(String, String, String)`, which again requires explicitly passing an id, type or source.

So, if all I want to do is set the CloudEvent subject on an outgoing message, and let Kalix set the remaining attributes - I can't.

**Comment by johanandren** [comment]
So, we deprecate, and add a `Metadata#withSubjectId(id)` or something like that?

**Comment by patriknw** [comment]
This is what I had to do:
```
    var metadata = actionContext().metadata().set("ce-subject", request.getTransactionId());

    return effects()
        .reply(topicTransaction, metadata);
```

**Comment by jroper** [comment]
But what about other CloudEvent metadata? There's all sorts of CloudEvent data users might want to set.

I'm not sure though if we want to pass all metadata along as is - technically I think these are different events and should have different cloud event IDs, plus the incoming source references the event source entity type, not something we probably want to expose publicly.

Also, the fact that we encode the CloudEvent subject as `ce-subject`, while consistent with most CloudEvent header based encodings, is supposed to be more of an implementation detail.

**Comment by johanandren** [comment]
Very problematic to pass along incoming cloud metadata to the response as it includes `ce-type` for the incoming message type which when returned in metadata overrides the type of the response message and leads to a downstream cloud event where metadata does not match the payload.

**Comment by jroper** [comment]
Yes, so this is why I think we should make it easy for users to just set a subset of CloudEvent metadata that they want to send, and allow other CloudEvent metadata, such as `ce-type`, be set by the proxy. The current API doesn't allow that, it forces the user to set a type, id and source.

**Comment by aludwiko** [comment]
I have a follow-up question. If I would like to integrate Kalix with an external system I can consume/produce raw bytes. What about raw attributes/metadata. As a user, I would assume that I can consume all the metadata and produce any metadata I need for the integration.

**Comment by jroper** [comment]
Nothing stopping you from doing that. Our CloudEvent metadata sits on top of the generic metadata, CloudEvent attributes get encoded with properties prepended with `ce-`, which is how it works in most CloudEvent encodings. But users are free to add whatever arbitrary properties they want.

**Comment by jroper** [comment]
You're right, that's a problem. We should pass all metadata along.

**Comment by johanandren** [comment]
In addition to making this API more useful we should also document it better, the current state is a test shown here: https://docs.kalix.io/java-protobuf/publishing-subscribing.html#_metadata but that doesn't really help with writing an action that emits events and wants to control cloud event metadata.
